src/FilenameParsing/FileIDHelperResolutionStrategy.php

```diff
@@ -432,31 +434,55 @@ public function findVariants($tuple, Filesystem $filesystem)
             // Make sure our yield file has an hash
             $hash = $parsedFileID->getHash() ?: $this->findHashOf($helper, $parsedFileID, $filesystem);
 
-
             // Find the correct folder to search for possible variants in
             $folder = $helper->lookForVariantIn($parsedFileID);
-            $possibleVariants = $filesystem->listContents($folder, $helper->lookForVariantRecursive())->toArray();
 
-            // Flysystem returns array of meta data abouch each file, we remove directories and map it down to the path
-            $possibleVariants = array_filter($possibleVariants ?? [], function ($possibleVariant) {
-                return $possibleVariant['type'] !== 'dir';
+            if (!$filesystem->directoryExists($folder)) {
+                continue;
+            }
+
+            // Use a glob if possible, as that will be more performant if the adapter supports it.
+            // We have to assume any helper we don't ship implements a different variant naming convention
+            // for backwards compatibility.
+            $helperAllowsGlob = in_array(get_class($helper), [
+                HashFileIDHelper::class,
+                NaturalFileIDHelper::class
+            ]);
+            if ($helperAllowsGlob && $filesystem instanceof GlobContentLister) {
```

Comment:
The `GlobContentLister` type check is necessary here because the `Filesystem` type in the method signature is `League\Flysystem\Filesystem`, not `SilverStripe\Assets\Flysystem\Filesystem`.